FIAT/polynomial_set.py

```diff
@@ -176,15 +176,57 @@ def polynomial_set_union_normalized(A, B):
     not contain any of the same members of the set, as we construct a
     span via SVD.
     """
-    new_coeffs = numpy.concatenate((A.coeffs, B.coeffs), axis=0)
+    assert A.get_reference_element() == B.get_reference_element()
+    new_coeffs = construct_new_coeffs(A.get_reference_element(), A, B)
+
+    deg = max(A.get_degree(), B.get_degree())
+    em_deg = max(A.get_embedded_degree(), B.get_embedded_degree())
     coeffs = spanning_basis(new_coeffs)
     return PolynomialSet(A.get_reference_element(),
-                         A.get_degree(),
-                         A.get_embedded_degree(),
+                         deg,
+                         em_deg,
                          A.get_expansion_set(),
                          coeffs)
 
 
+def construct_new_coeffs(ref_el, A, B):
+    """
+    Constructs new coefficients for the set union of A and B
+    If A and B are discontinuous and do not have the same degree the smaller one
+    is extended to match the larger.
+
+    This does not handle the case that A and B have continuity but not the same degree.
+    """
+
+    sd = ref_el.get_spatial_dimension()
+    if A.get_expansion_set().continuity != B.get_expansion_set().continuity:
+        raise ValueError("Continuity of expansion sets does not match.")
+    if A.get_embedded_degree() != B.get_embedded_degree() and A.get_expansion_set().continuity is None:
+        higher = A if A.get_embedded_degree() > B.get_embedded_degree() else B
+        lower = B if A.get_embedded_degree() > B.get_embedded_degree() else A
+
+        try:
+            sd = lower.get_shape()[0]
+        except IndexError:
+            sd = -1
+
+        embedded_coeffs = []
+        diff = higher.coeffs.shape[-1] - lower.coeffs.shape[-1]
+        for coeff in lower.coeffs:
+            if sd != -1:
+                new_coeff = []
+                for row in coeff:
+                    new_coeff.append(numpy.append(row, [0 for i in range(diff)]))
+                embedded_coeffs.append(new_coeff)
+            else:
+                embedded_coeffs.append(numpy.append(coeff, [0 for i in range(diff)]))
+        embedded_coeffs = numpy.array(embedded_coeffs)
```

Comment:
Could this be done by first creating `embedded_coeffs` as `numpy.zeros` of the right shape and then inserting the `lower.coeffs` on the right slice? Alternatively, I think one could use  `numpy.pad` here.

Reply:
Maybe it is something like 

```
size = lower.coeffs.shape[-1]
embedded_coeffs = numpy.zeros(higher.coeffs.shape)
embedded_coeffs[..., :size] = lower.coeffs
```

Reply:
have changed it to use np.pad- thanks!